File: hooks.ts
// hooks.ts — Phase lifecycle hook execution via execFileSync

import { execFileSync } from 'node:child_process'
import { readFileSync } from 'node:fs'
import { resolve, normalize, isAbsolute, join } from 'node:path'
import type { AgentDirective, HookConfig, HookTrigger, PipelineRunParameters } from './types.ts'
import { PipelineError } from './types.ts'

/** Context passed to hook commands via PIPELINE_HOOK_CONTEXT env var. */
export interface HookContext {
  run_id: string
  phase: string
  trigger: HookTrigger
  project_root: string
  run_dir: string
  [key: string]: unknown
}

/**
 * Context passed to `pre_pipeline_init` hooks via PIPELINE_HOOK_CONTEXT env var.
 * Distinct from `HookContext` because there is no phase or run yet — this hook
 * fires BEFORE `pipeline_init_run` mutates state. The hook can inspect the
 * raw tool arguments and emit parameters or user prompts.
 */
export interface PreInitHookContext {
  trigger: 'pre_pipeline_init'
  project_root: string
  /** Raw `pipeline_init_run` tool arguments (e.g. run_id, skip_phases, etc.). */
  requested_args: Record<string, unknown>
}

/** Result of running all `pre_pipeline_init` hooks: merged parameters plus any user prompts. */
export interface PreInitHookResult {
  parameters: PipelineRunParameters
  userPrompts: string[]
}

/** Result of running a single hook. */
export interface HookResult {
  command: string
  trigger: HookTrigger
  phase: string
  success: boolean
  stdout?: string
  stderr?: string
  error?: string
  /**
   * Feature B directive parsed from the hook's JSON stdout. Present only
   * when the hook emitted a JSON object containing a well-formed
   * `agent_directive` with all four required string fields
   * (`subagent_type`, `model`, `description`, `prompt`). Feature B's
   * `handleStartPhase` forwards this into the `pipeline_start_phase` MCP
   * tool response so the client can spawn the phase subagent.
   */
  agent_directive?: AgentDirective
}

/**
 * Parse an `AgentDirective` from a hook's raw stdout string.
 *
 * The stdout must be a JSON object literal containing an `agent_directive`
 * field whose value has all four required string fields (`subagent_type`,
 * `model`, `description`, `prompt`). An optional `isolation: 'worktree'`
 * is preserved when present; any other `isolation` value is dropped.
 *
 * Returns `undefined` when the stdout is absent, empty, not JSON, missing
 * the `agent_directive` field, or the directive fails shape validation.
 * This helper is intentionally lenient so existing non-directive hooks
 * (plain text output, JSON without a directive) pass through unchanged —
 * only Feature B `pre_start` hooks that opt in by emitting the correctly
 * shaped JSON attach a directive to the `HookResult`.
 */
export function parseAgentDirective(stdout: string | undefined): AgentDirective | undefined {
  if (!stdout) return undefined
  const trimmed = stdout.trim()
  if (trimmed.length === 0) return undefined

  let parsed: unknown
  try {
    parsed = JSON.parse(trimmed)
  } catch {
    return undefined
  }

  if (!parsed || typeof parsed !== 'object') return undefined
  const root = parsed as Record<string, unknown>
  const raw = root.agent_directive
  if (!raw || typeof raw !== 'object') return undefined

  const d = raw as Record<string, unknown>
  const { subagent_type, model, description, prompt } = d
  if (
    typeof subagent_type !== 'string' ||
    typeof model !== 'string' ||
    typeof description !== 'string' ||
    typeof prompt !== 'string'
  ) {
    return undefined
  }

  const directive: AgentDirective = { subagent_type, model, description, prompt }
  if (d.isolation === 'worktree') {
    directive.isolation = 'worktree'
  }
  return directive
}

const MAX_CONTEXT_BYTES = 65536 // 64KB cap on env var context

/**
 * Validate that a hook command path is safe:
 * - Resolve against project root
 * - Reject any path containing '..' after normalization
 * - Must resolve to a path under project root
 */
export function validateHookPath(command: string, projectRoot: string): string {
  const normalized = normalize(command)

  // Reject paths with '..' traversal after normalization
  if (normalized.includes('..')) {
    throw new PipelineError(
      `Hook command path "${command}" contains directory traversal (..)`,
      'configuration_error',
      {
        recovery_action: 'Use a path relative to the project root without ".." components.',
        details: { command, normalized },
      },
    )
  }

  // Resolve against project root
  const resolved = isAbsolute(normalized) ? normalized : resolve(projectRoot, normalized)

  // Verify resolved path is under project root
  const normalizedRoot = normalize(projectRoot)
  if (!resolved.startsWith(normalizedRoot)) {
    throw new PipelineError(
      `Hook command path "${command}" resolves outside project root`,
      'configuration_error',
      {
        recovery_action: 'Ensure hook command path is within the project directory.',
        details: { command, resolved, projectRoot },
      },
    )
  }

  return resolved
}

/**
 * Check if a hook's phase_filter matches the given phase name.
 * Supports '*' (all phases) or exact match.
 */
export function matchesPhase(filter: string | undefined, phaseName: string): boolean {
  if (!filter || filter === '*') return true
  return filter === phaseName
}

/**
 * Check whether events.jsonl in the given run directory already contains a
 * `phase_completed` event for the target phase. Returns true if at least one
 * matching event is present, false otherwise (including when the file does not
 * exist or cannot be read).
 *
 * Used by runHooks to enforce idempotency for post_complete hooks: if the
 * event is already recorded, we know the hook has already fired once and
 * should not fire again.
 */
function hasPhaseCompletedEvent(runDir: string, phase: string): boolean {
  try {
    const eventsPath = join(runDir, 'events.jsonl')
    const content = readFileSync(eventsPath, 'utf-8')
    const lines = content.split('\n')
    for (const line of lines) {
      const trimmed = line.trim()
      if (!trimmed) continue
      try {
        const parsed = JSON.parse(trimmed) as { event?: unknown; phase?: unknown }
        if (parsed.event === 'phase_completed' && parsed.phase === phase) {
          return true
        }
      } catch {
        // Malformed line — skip defensively.
      }
    }
    return false
  } catch {
    // File missing or unreadable — treat as "no prior event".
    return false
  }
}

/**
 * Run hooks matching a specific trigger and phase.
 *
 * Pre-start hooks are blocking: if any hook fails, a PipelineError is thrown.
 * Post-complete and on_fail hooks are non-blocking: failures are returned as warnings.
 *
 * @param hooks - Array of hook configurations
 * @param trigger - Which trigger point to run
 * @param context - Hook context (passed as PIPELINE_HOOK_CONTEXT env var)
 * @param projectRoot - Project root for path resolution
 * @returns Array of hook results
 */
export function runHooks(
  hooks: HookConfig[],
  trigger: HookTrigger,
  context: HookContext,
  projectRoot: string,
): HookResult[] {
  const matching = hooks.filter(
    h => h.trigger === trigger && matchesPhase(h.phase_filter, context.phase),
  )

  if (matching.length === 0) return []

  // Phase idempotency for post_complete: if a phase_completed event already
  // exists in events.jsonl for this phase, skip all post_complete hooks for
  // this call and return a synthetic skipped result per matching hook.
  if (trigger === 'post_complete' && hasPhaseCompletedEvent(context.run_dir, context.phase)) {
    const timestamp = new Date().toISOString()
    // Stderr debug line for post-mortem investigation of hook-fire sequences.
    console.warn(
      `[hooks] SKIPPED post_complete for phase="${context.phase}" at ${timestamp} ` +
      `(events.jsonl already contains phase_completed for this phase)`,
    )
    return matching.map(hook => ({
      command: hook.command,
      trigger,
      phase: context.phase,
      success: false,
      error: `skipped: phase_completed already recorded for phase "${context.phase}" (idempotency guard)`,
    }))
  }

  // Debug logging keyed on trigger + phase + timestamp for post-mortem traces.
  console.warn(
    `[hooks] firing trigger="${trigger}" phase="${context.phase}" ` +
    `matched=${matching.length} at ${new Date().toISOString()}`,
  )

  const results: HookResult[] = []
  const contextJson = JSON.stringify(context)

  // Cap context size for env var safety
  const envContext = contextJson.length <= MAX_CONTEXT_BYTES
    ? contextJson
    : JSON.stringify({ error: 'context_too_large', run_id: context.run_id, phase: context.phase })

  for (const hook of matching) {
    let resolvedCommand: string
    try {
      resolvedCommand = validateHookPath(hook.command, projectRoot)
    } catch (err) {
      const result: HookResult = {
        command: hook.command,
        trigger,
        phase: context.phase,
        success: false,
        error: err instanceof Error ? err.message : String(err),
      }

      if (trigger === 'pre_start') {
        throw err // Re-throw for blocking hooks
      }

      results.push(result)
      continue
    }

    try {
      const stdout = execFileSync(resolvedCommand, hook.args ?? [], {
        timeout: hook.timeout_ms ?? 5000,
        env: { ...process.env, PIPELINE_HOOK_CONTEXT: envContext },
        encoding: 'utf-8',
        stdio: ['pipe', 'pipe', 'pipe'],
      })

      const trimmedStdout = typeof stdout === 'string' ? stdout.trim() : undefined
      const directive = parseAgentDirective(trimmedStdout)
      const result: HookResult = {
        command: hook.command,
        trigger,
        phase: context.phase,
        success: true,
        stdout: trimmedStdout,
      }
      if (directive) result.agent_directive = directive
      results.push(result)
    } catch (err: unknown) {
      const error = err instanceof Error ? err : new Error(String(err))
      const result: HookResult = {
        command: hook.command,
        trigger,
        phase: context.phase,
        success: false,
        stderr: (error as { stderr?: string }).stderr,
        error: error.message,
      }

      if (trigger === 'pre_start') {
        // Blocking: throw PipelineError
        throw new PipelineError(
          `Pre-start hook failed for phase "${context.phase}": ${hook.command} — ${error.message}`,
          'configuration_error',
          {
            recovery_action: 'Fix the hook script or remove it from hooks configuration.',
            details: {
              command: hook.command,
              phase: context.phase,
              error: error.message,
            },
          },
        )
      }

      // Non-blocking: record and continue
      results.push(result)
    }
  }

  return results
}

/**
 * Run all `pre_pipeline_init` hooks in order. Each hook's stdout must be a
 * JSON object shaped `{ parameters?: PipelineRunParameters, userPrompts?: string[] }`.
 * Parameters from later hooks override earlier ones; prompts accumulate.
 *
 * Unlike phase hooks, this trigger has no phase filter — every enabled hook
 * with `trigger: 'pre_pipeline_init'` fires.
 *
 * This function is blocking: if any hook script fails (non-zero exit, timeout,
 * malformed JSON, path traversal), a `PipelineError('configuration_error')` is thrown.
 * Returns `{ parameters: {}, userPrompts: [] }` when no matching hooks exist.
 *
 * @param hooks - Array of hook configurations (filtered from the loaded `[[hooks]]` table)
 * @param context - Pre-init hook context (passed as PIPELINE_HOOK_CONTEXT env var)
 * @param projectRoot - Project root for path resolution via validateHookPath
 */
export function runPrePipelineInitHooks(
  hooks: HookConfig[],
  context: PreInitHookContext,
  projectRoot: string,
): PreInitHookResult {
  const matching = hooks.filter(h => h.trigger === 'pre_pipeline_init')
  const merged: PipelineRunParameters = {}
  const allPrompts: string[] = []

  if (matching.length === 0) {
    return { parameters: merged, userPrompts: allPrompts }
  }

  console.warn(
    `[hooks] firing trigger="pre_pipeline_init" matched=${matching.length} at ${new Date().toISOString()}`,
  )

  const contextJson = JSON.stringify(context)
  const envContext = contextJson.length <= MAX_CONTEXT_BYTES
    ? contextJson
    : JSON.stringify({ error: 'context_too_large', trigger: 'pre_pipeline_init' })

  for (const hook of matching) {
    let resolvedCommand: string
    try {
      resolvedCommand = validateHookPath(hook.command, projectRoot)
    } catch (err) {
      // Re-throw path-traversal / outside-root errors as configuration_error
      throw err
    }

    try {
      const stdout = execFileSync(resolvedCommand, hook.args ?? [], {
        timeout: hook.timeout_ms ?? 5000,
        env: { ...process.env, PIPELINE_HOOK_CONTEXT: envContext },
        encoding: 'utf-8',
        stdio: ['pipe', 'pipe', 'pipe'],
      })
      const trimmed = typeof stdout === 'string' ? stdout.trim() : ''
      if (trimmed.length === 0) continue

      let parsed: { parameters?: unknown; userPrompts?: unknown; user_prompts?: unknown; user_prompt?: unknown }
      try {
        parsed = JSON.parse(trimmed)
      } catch (jsonErr) {
        throw new PipelineError(
          `pre_pipeline_init hook "${hook.command}" emitted non-JSON stdout: ${jsonErr instanceof Error ? jsonErr.message : String(jsonErr)}`,
          'configuration_error',
          {
            recovery_action: 'Ensure the hook script writes a JSON object to stdout with { parameters, userPrompts } shape.',
            details: { command: hook.command, stdout_preview: trimmed.slice(0, 200) },
          },
        )
      }

      if (parsed.parameters && typeof parsed.parameters === 'object') {
        Object.assign(merged, parsed.parameters as PipelineRunParameters)
      }
      if (Array.isArray(parsed.userPrompts)) {
        for (const p of parsed.userPrompts) {
          if (typeof p === 'string') allPrompts.push(p)
        }
      }
      if (Array.isArray(parsed.user_prompts)) {
        for (const p of parsed.user_prompts) {
          if (typeof p === 'string') allPrompts.push(p)
        }
      }
      if (typeof parsed.user_prompt === 'string') {
        allPrompts.push(parsed.user_prompt)
      }
    } catch (err: unknown) {
      if (err instanceof PipelineError) throw err
      const error = err instanceof Error ? err : new Error(String(err))
      throw new PipelineError(
        `pre_pipeline_init hook failed: ${hook.command} — ${error.message}`,
        'configuration_error',
        {
          recovery_action: 'Fix the hook script or remove it from the [[hooks]] configuration.',
          details: { command: hook.command, error: error.message },
        },
      )
    }
  }

  return { parameters: merged, userPrompts: allPrompts }
}

File: src/pipeline_orchestrator/hooks.py
"""Phase lifecycle hook execution seam (port of ``legacy-node/hooks.ts``, spec §7.6).

Byte-exact behavioral port of the Node ``hooks.ts`` module: the subprocess
runner that fires lifecycle hooks via :func:`subprocess.run`, the path-traversal
guard, the phase-filter matcher, the ``agent_directive`` parser, the
``pre_pipeline_init`` parameter/prompt collector, and the ``post_complete``
idempotency guard keyed on ``events.jsonl``.

Parity notes carried over from the Node source:

* Hook context is passed via the ``PIPELINE_HOOK_CONTEXT`` env var, capped at
  64KB; an oversized context is replaced with a small ``context_too_large``
  marker object (NOT truncated) — preserving the Node ``MAX_CONTEXT_BYTES`` cap.
* ``timeout_ms`` defaults to ``5000``; the default is applied with
  ``.get("timeout_ms", 5000)`` semantics (the TS ``?? 5000``) so an explicit
  ``0`` would survive — but :class:`~pipeline_orchestrator.models.HookConfig`
  already fills ``5000`` via the loader.
* ``pre_start`` hooks are blocking (failures re-raise as a
  ``PipelineError('configuration_error')``); ``post_complete`` / ``on_fail``
  hooks are non-blocking (failures recorded as ``success: False`` results).
* ``runPrePipelineInitHooks`` is blocking; later hooks' ``parameters`` override
  earlier ones (``dict.update`` = ``Object.assign`` later-wins) and ``userPrompts``
  accumulate, including the legacy ``user_prompts`` / ``user_prompt`` aliases.
* Diagnostics go to **stderr** (the Node ``console.warn``) — never stdout, which
  would corrupt the JSON-RPC channel.

The :class:`HookResult` / :class:`PreInitHookResult` returns are plain dicts
keyed exactly as the Node object literals, and a parsed ``agent_directive`` is a
plain dict with the four required string fields (plus an optional
``isolation: 'worktree'``) — matching what the Node tests ``deepEqual`` against.
"""

from __future__ import annotations

import json
import os
import subprocess
import sys
from dataclasses import dataclass, field
from datetime import UTC, datetime
from typing import Any

from .errors import ErrorClass, PipelineError
from .models import HookConfig, HookTrigger

#: 64KB cap on the ``PIPELINE_HOOK_CONTEXT`` env var (the Node ``MAX_CONTEXT_BYTES``).
MAX_CONTEXT_BYTES = 65536


@dataclass
class PreInitHookResult:
    """Result of running all ``pre_pipeline_init`` hooks (TS ``PreInitHookResult``).

    ``parameters`` is the later-wins merge of every hook's emitted ``parameters``;
    ``userPrompts`` accumulates ``userPrompts`` / ``user_prompts`` / ``user_prompt``
    across hooks. Field names match the TS object so callers/tests read identically.
    """

    parameters: dict[str, Any] = field(default_factory=dict)
    userPrompts: list[str] = field(default_factory=list)  # noqa: N815


def _now_iso() -> str:
    """Return the current UTC time as an ISO 8601 string (Node ``toISOString()``)."""
    now = datetime.now(UTC)
    return now.strftime("%Y-%m-%dT%H:%M:%S.") + f"{now.microsecond // 1000:03d}Z"


def parse_agent_directive(stdout: str | None) -> dict[str, Any] | None:
    """Parse an ``AgentDirective`` from a hook's raw stdout string.

    Mirrors ``parseAgentDirective`` in ``hooks.ts``. The stdout must be a JSON
    object containing an ``agent_directive`` field whose value has all four
    required string fields (``subagent_type`` / ``model`` / ``description`` /
    ``prompt``). An optional ``isolation: 'worktree'`` is preserved; any other
    ``isolation`` value is dropped. Returns ``None`` when stdout is absent, empty,
    not JSON, missing the field, or fails shape validation. Intentionally lenient
    so non-directive hooks pass through unchanged.
    """
    if not stdout:
        return None
    trimmed = stdout.strip()
    if len(trimmed) == 0:
        return None

    try:
        parsed: Any = json.loads(trimmed)
    except (json.JSONDecodeError, ValueError):
        return None

    if not parsed or not isinstance(parsed, dict):
        return None
    raw = parsed.get("agent_directive")
    if not raw or not isinstance(raw, dict):
        return None

    subagent_type = raw.get("subagent_type")
    model = raw.get("model")
    description = raw.get("description")
    prompt = raw.get("prompt")
    if (
        not isinstance(subagent_type, str)
        or not isinstance(model, str)
        or not isinstance(description, str)
        or not isinstance(prompt, str)
    ):
        return None

    directive: dict[str, Any] = {
        "subagent_type": subagent_type,
        "model": model,
        "description": description,
        "prompt": prompt,
    }
    if raw.get("isolation") == "worktree":
        directive["isolation"] = "worktree"
    return directive


def validate_hook_path(command: str, project_root: str) -> str:
    """Validate that a hook command path is safe and resolve it under ``project_root``.

    Mirrors ``validateHookPath`` in ``hooks.ts``:

    * Normalize the command path.
    * Reject any path containing ``..`` after normalization
      (``configuration_error``).
    * Resolve relative paths against ``project_root``; leave absolute paths as-is.
    * Verify the resolved path begins with the normalized project root
      (``configuration_error`` otherwise).

    Uses ``os.path.normpath`` (the Node ``path.normalize`` analogue) and
    ``os.path.isabs`` / ``os.path.join`` so behavior matches per-platform path
    semantics, exactly like the Node source.
    """
    normalized = os.path.normpath(command)

    # Reject paths with '..' traversal after normalization. Mirrors the Node
    # ``normalized.includes('..')`` — a substring check (intentionally broad:
    # it also rejects a filename that merely contains ".." as a substring),
    # preserved verbatim for byte-exact parity with hooks.ts.
    if ".." in normalized:
        raise PipelineError(
            f'Hook command path "{command}" contains directory traversal (..)',
            ErrorClass.configuration_error,
            recovery_action=(
                'Use a path relative to the project root without ".." components.'
            ),
            details={"command": command, "normalized": normalized},
        )

    # Resolve against project root.
    resolved = (
        normalized
        if os.path.isabs(normalized)
        else os.path.normpath(os.path.join(project_root, normalized))
    )

    # Verify resolved path is under project root.
    normalized_root = os.path.normpath(project_root)
    if not resolved.startswith(normalized_root):
        raise PipelineError(
            f'Hook command path "{command}" resolves outside project root',
            ErrorClass.configuration_error,
            recovery_action="Ensure hook command path is within the project directory.",
            details={
                "command": command,
                "resolved": resolved,
                "projectRoot": project_root,
            },
        )

    return resolved


def matches_phase(filter_: str | None, phase_name: str) -> bool:
    """Check whether a hook's ``phase_filter`` matches the given phase name.

    Mirrors ``matchesPhase`` in ``hooks.ts``: ``'*'`` (or absent) matches all
    phases; otherwise an exact string match is required.
    """
    if not filter_ or filter_ == "*":
        return True
    return filter_ == phase_name


def _has_phase_completed_event(run_dir: str, phase: str) -> bool:
    """Return whether ``events.jsonl`` already has a ``phase_completed`` for ``phase``.

    Mirrors ``hasPhaseCompletedEvent`` in ``hooks.ts``. Returns ``False`` when the
    file is missing/unreadable or no matching line is present; malformed JSON
    lines are skipped defensively.
    """
    try:
        events_path = os.path.join(run_dir, "events.jsonl")
        with open(events_path, encoding="utf-8") as fh:
            content = fh.read()
    except OSError:
        # File missing or unreadable — treat as "no prior event".
        return False

    for line in content.split("\n"):
        trimmed = line.strip()
        if not trimmed:
            continue
        try:
            parsed = json.loads(trimmed)
        except (json.JSONDecodeError, ValueError):
            # Malformed line — skip defensively.
            continue
        if (
            isinstance(parsed, dict)
            and parsed.get("event") == "phase_completed"
            and parsed.get("phase") == phase
        ):
            return True
    return False


def _execute_hook(
    resolved_command: str, args: list[str], timeout_ms: int, env_context: str
) -> str:
    """Run a hook subprocess and return its stdout, raising on failure/timeout.

    Mirrors the Node ``execFileSync(..., { timeout, env, encoding, stdio })``:
    inherits the process env with ``PIPELINE_HOOK_CONTEXT`` overlaid, captures
    stdout/stderr as text, and raises :class:`subprocess.CalledProcessError`
    (non-zero exit, carrying ``.stderr``) or :class:`subprocess.TimeoutExpired`.
    """
    env = {**os.environ, "PIPELINE_HOOK_CONTEXT": env_context}
    completed = subprocess.run(  # noqa: S603
        [resolved_command, *args],
        capture_output=True,
        text=True,
        timeout=timeout_ms / 1000,
        env=env,
        check=True,
    )
    return completed.stdout


def run_hooks(
    hooks: list[HookConfig],
    trigger: HookTrigger,
    context: dict[str, Any],
    project_root: str,
) -> list[dict[str, Any]]:
    """Run hooks matching a specific trigger and phase.

    Mirrors ``runHooks`` in ``hooks.ts``. ``pre_start`` hooks are blocking (a
    failure re-raises a ``PipelineError('configuration_error')``);
    ``post_complete`` / ``on_fail`` hooks are non-blocking (failures recorded as
    ``success: False`` results). For ``post_complete``, an idempotency guard short
    -circuits when ``events.jsonl`` already records a ``phase_completed`` for the
    target phase, returning a synthetic skipped result per matching hook.

    ``context`` is the ``HookContext`` dict (``run_id`` / ``phase`` / ``trigger`` /
    ``project_root`` / ``run_dir`` plus any extra keys); each returned result is a
    plain dict keyed exactly as the Node ``HookResult`` object literal.
    """
    phase = context["phase"]
    matching = [
        h
        for h in hooks
        if h.trigger == trigger and matches_phase(h.phase_filter, phase)
    ]

    if len(matching) == 0:
        return []

    # Phase idempotency for post_complete.
    if trigger == "post_complete" and _has_phase_completed_event(
        context["run_dir"], phase
    ):
        timestamp = _now_iso()
        # Stderr debug line for post-mortem investigation of hook-fire sequences.
        print(
            f'[hooks] SKIPPED post_complete for phase="{phase}" at {timestamp} '
            "(events.jsonl already contains phase_completed for this phase)",
            file=sys.stderr,
        )
        return [
            {
                "command": hook.command,
                "trigger": trigger,
                "phase": phase,
                "success": False,
                "error": (
                    "skipped: phase_completed already recorded for phase "
                    f'"{phase}" (idempotency guard)'
                ),
            }
            for hook in matching
        ]

    # Debug logging keyed on trigger + phase + timestamp for post-mortem traces.
    print(
        f'[hooks] firing trigger="{trigger}" phase="{phase}" '
        f"matched={len(matching)} at {_now_iso()}",
        file=sys.stderr,
    )

    results: list[dict[str, Any]] = []
    context_json = json.dumps(context, separators=(",", ":"))

    # Cap context size for env var safety.
    if len(context_json) <= MAX_CONTEXT_BYTES:
        env_context = context_json
    else:
        env_context = json.dumps(
            {
                "error": "context_too_large",
                "run_id": context["run_id"],
                "phase": phase,
            },
            separators=(",", ":"),
        )

    for hook in matching:
        try:
            resolved_command = validate_hook_path(hook.command, project_root)
        except PipelineError as err:
            result: dict[str, Any] = {
                "command": hook.command,
                "trigger": trigger,
                "phase": phase,
                "success": False,
                "error": err.message,
            }
            if trigger == "pre_start":
                raise  # Re-throw for blocking hooks.
            results.append(result)
            continue

        try:
            stdout = _execute_hook(
                resolved_command,
                hook.args,
                hook.timeout_ms,
                env_context,
            )
            trimmed_stdout = stdout.strip() if isinstance(stdout, str) else None
            directive = parse_agent_directive(trimmed_stdout)
            ok_result: dict[str, Any] = {
                "command": hook.command,
                "trigger": trigger,
                "phase": phase,
                "success": True,
                "stdout": trimmed_stdout,
            }
            if directive:
                ok_result["agent_directive"] = directive
            results.append(ok_result)
        except (
            subprocess.CalledProcessError,
            subprocess.TimeoutExpired,
            OSError,
        ) as err:
            message = str(err)
            stderr = getattr(err, "stderr", None)
            err_result: dict[str, Any] = {
                "command": hook.command,
                "trigger": trigger,
                "phase": phase,
                "success": False,
                "stderr": stderr,
                "error": message,
            }

            if trigger == "pre_start":
                # Blocking: throw PipelineError.
                raise PipelineError(
                    f'Pre-start hook failed for phase "{phase}": '
                    f"{hook.command} — {message}",
                    ErrorClass.configuration_error,
                    recovery_action=(
                        "Fix the hook script or remove it from hooks configuration."
                    ),
                    details={
                        "command": hook.command,
                        "phase": phase,
                        "error": message,
                    },
                ) from err

            # Non-blocking: record and continue.
            results.append(err_result)

    return results


def run_pre_pipeline_init_hooks(
    hooks: list[HookConfig],
    context: dict[str, Any],
    project_root: str,
) -> PreInitHookResult:
    """Run all ``pre_pipeline_init`` hooks in order, merging parameters and prompts.

    Mirrors ``runPrePipelineInitHooks`` in ``hooks.ts``. Each hook's stdout must
    be a JSON object shaped ``{ parameters?, userPrompts? }`` (plus legacy
    ``user_prompts`` / ``user_prompt`` aliases). Later hooks' ``parameters``
    override earlier ones; prompts accumulate. Blocking: any hook failure
    (non-zero exit, timeout, malformed JSON, path traversal) raises a
    ``PipelineError('configuration_error')``. Returns an empty result when no
    ``pre_pipeline_init`` hooks exist.
    """
    matching = [h for h in hooks if h.trigger == "pre_pipeline_init"]
    merged: dict[str, Any] = {}
    all_prompts: list[str] = []

    if len(matching) == 0:
        return PreInitHookResult(parameters=merged, userPrompts=all_prompts)

    print(
        f'[hooks] firing trigger="pre_pipeline_init" '
        f"matched={len(matching)} at {_now_iso()}",
        file=sys.stderr,
    )

    context_json = json.dumps(context, separators=(",", ":"))
    if len(context_json) <= MAX_CONTEXT_BYTES:
        env_context = context_json
    else:
        env_context = json.dumps(
            {"error": "context_too_large", "trigger": "pre_pipeline_init"},
            separators=(",", ":"),
        )

    for hook in matching:
        # Re-throw path-traversal / outside-root errors as configuration_error.
        resolved_command = validate_hook_path(hook.command, project_root)

        try:
            stdout = _execute_hook(
                resolved_command,
                hook.args,
                hook.timeout_ms,
                env_context,
            )
        except (
            subprocess.CalledProcessError,
            subprocess.TimeoutExpired,
            OSError,
        ) as err:
            message = str(err)
            raise PipelineError(
                f"pre_pipeline_init hook failed: {hook.command} — {message}",
                ErrorClass.configuration_error,
                recovery_action=(
                    "Fix the hook script or remove it from the [[hooks]] configuration."
                ),
                details={"command": hook.command, "error": message},
            ) from err

        trimmed = stdout.strip() if isinstance(stdout, str) else ""
        if len(trimmed) == 0:
            continue

        try:
            parsed: Any = json.loads(trimmed)
        except (json.JSONDecodeError, ValueError) as json_err:
            raise PipelineError(
                f'pre_pipeline_init hook "{hook.command}" emitted non-JSON stdout: '
                f"{json_err}",
                ErrorClass.configuration_error,
                recovery_action=(
                    "Ensure the hook script writes a JSON object to stdout with "
                    "{ parameters, userPrompts } shape."
                ),
                details={
                    "command": hook.command,
                    "stdout_preview": trimmed[:200],
                },
            ) from json_err

        if not isinstance(parsed, dict):
            parsed = {}

        params = parsed.get("parameters")
        if params and isinstance(params, dict):
            merged.update(params)
        user_prompts = parsed.get("userPrompts")
        if isinstance(user_prompts, list):
            for p in user_prompts:
                if isinstance(p, str):
                    all_prompts.append(p)
        legacy_prompts = parsed.get("user_prompts")
        if isinstance(legacy_prompts, list):
            for p in legacy_prompts:
                if isinstance(p, str):
                    all_prompts.append(p)
        legacy_single = parsed.get("user_prompt")
        if isinstance(legacy_single, str):
            all_prompts.append(legacy_single)

    return PreInitHookResult(parameters=merged, userPrompts=all_prompts)
